x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyAuthenticator.java

```diff
@@ -54,13 +57,15 @@ public void authenticate(Context context, ActionListener<AuthenticationResult<Au
         apiKeyService.tryAuthenticate(context.getThreadContext(), apiKeyCredentials, ActionListener.wrap(authResult -> {
```

Comment:
I've also considered an option to wrap the action listener in order to keep the code simpler and avoid the redundant calls to `recordFailedAuthentication`. But this did not feel like a strong enough reason, compared to an overhead and impact to performance by wrapping yet another listener.


```java
ActionListener.wrap(authResult -> {
            if (authResult.isAuthenticated()) {
                recordSuccessfulAuthentication(apiKeyCredentials);
            } else {
                recordFailedAuthentication(apiKeyCredentials, authResult.getMessage());
            }
            listener.onResponse(authResult);
        }, e -> {
            recordFailedAuthentication(apiKeyCredentials, e.getMessage());
            listener.onFailure(e);
        });
```